### PR TITLE
Charge gas for local variables on the callee side

### PR DESCRIPTION
**Pull request by agryaznov**

### CODE REVIEW

**Comment by athei** [comment]
Please bump minor version and add README entry.

File: CHANGELOG.md [hunk 1]
```diff
@@ -22,6 +22,7 @@ In other words: Upgrading this pallet will not break pre-existing contracts.
 
 - Add new gas metering method: mutable global + local gas function
 [#34](https://github.com/paritytech/wasm-instrument/pull/34)
+- Account for locals initialization costs [#38](https://github.com/paritytech/wasm-instrument/pull/38)
```
Comment (athei) [inline CHANGELOG.md:25]:
```suggestion
- Account for locals initialization costs
[#38](https://github.com/paritytech/wasm-instrument/pull/38)
```

**Review by athei — APPROVED** [review]

File: src/gas_metering/validation.rs [hunk 2]
```diff
@@ -134,6 +134,10 @@ fn build_control_flow_graph(
 
 	let mut stack = vec![ControlFrame::new(entry_node_id, terminal_node_id, false)];
 	let mut metered_blocks_iter = blocks.iter().peekable();
+
+	let locals_count = body.locals().iter().fold(0, |count, val_type| count + val_type.count());
```
Comment (pepyakin) [inline src/gas_metering/validation.rs:138]: can't the + overflow there?

File: src/gas_metering/mod.rs [hunk 3]
```diff
@@ -249,8 +257,16 @@ pub fn inject<R: Rules, B: Backend>(
 							}
 						}
 					}
-					if inject_counter(func_body.code_mut(), gas_fn_cost, rules, gas_func_idx)
-						.is_err()
+					let locals_count =
+						func_body.locals().iter().map(|val_type| val_type.count()).sum();
```
Comment (pepyakin) [inline src/gas_metering/mod.rs:261]: can't `sum` overflow here?

File: src/gas_metering/validation.rs (hunk 2 above)
Comment (athei) [reply src/gas_metering/validation.rs:138]: I don't think so. Wasm standard limits number of locals to 50k. Not sure if `parity-wasm` enforces that, though. We should probably work with checked math here.

File: src/gas_metering/mod.rs (hunk 3 above)
Comment (athei) [reply src/gas_metering/mod.rs:261]: Same as above.

File: src/gas_metering/validation.rs (hunk 2 above)
Comment (pepyakin) [reply src/gas_metering/validation.rs:138]: I certainly am not aware of any such constraints! I do know that implementations of the spec could impose structural limits, but those are just expected to be reasonable for some definition of that word.

Comment (agryaznov) [reply src/gas_metering/validation.rs:138]: addressed in the follow-up PR

Comment (pepyakin) [reply src/gas_metering/validation.rs:138]: fwiw the relevant [part](https://webassembly.github.io/spec/core/appendix/implementation.html#syntactic-limits) of the spec

Comment (athei) [reply src/gas_metering/validation.rs:138]: Okay then it was only a wasmparser specific implementation limitation.

**Pull request merged**